Add UpgradeOnMovement trait

reaperrr opened the pull request:
As the name states, the trait provides upgrades while an actor is moving, this can be used to increase inaccuracy while moving, give the unit a more vulnerable armor type while moving and so on.

This nicely complements our set of upgrades-providing traits.

The trait is upgradable itself, which could be used to switch the provided upgrades manually in conjunction with `DeployToUpgrade`.

reaperrr committed:
Add UpgradeOnMovement trait

This provides upgrades when an actor is moving. To avoid 'hiccups' for ground units, air units are considered moving as soon as they are airborne.

ABrandau commented:
Seems to work fine, I emped the unit to see if it had any issues and nope, same as when the unit attack moves.

WolfGaming commented:
Do we want to have a property that flips the functionality? Aka grants an upgrade when not moving?

pchote commented:
The upgrades system already lets you do that without making a new type.

WolfGaming commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs:
```diff
@@ -0,0 +1,68 @@
+		readonly IMove movement;
+
+		bool granted;
+		bool isMoving;
```
Why have this here and not defined in the tick like oldCachedPosition?

WolfGaming commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs:
```diff
@@ -0,0 +1,68 @@
+			var oldCachedPosition = cachedPosition;
+			cachedPosition = self.CenterPosition;
+
+			isMoving = !self.IsDead && (movement.IsMoving || (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0);
```
Do you do (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0 for actors inside transports or am I missing something?

reaperrr replied:
During my tests there were some 'hiccups' when only using IMove.IsMoving (during turns, usually). If we managed to make that more robust instead this could go away, but until then this is needed to make sure the upgrades are applied all the time while moving.

reaperrr commented:
Updated:
- upgrades are now also revoked if `UpgradeOnMovement` is disabled
- `isMovement` is now a `var` inside Tick
- moved the testcase to https://github.com/reaperrr/OpenRA/tree/UpgradeOnMovement-test

abcdefg30 commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs:
```diff
@@ -0,0 +1,76 @@
+			var oldCachedPosition = cachedPosition;
+			cachedPosition = self.CenterPosition;
+
+			var isMoving = !self.IsDead && (movement.IsMoving || (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0);
```
This "breaks" with aircraft. They always have `movement.IsMoving = self.World.Map.DistanceAboveTerrain(CenterPosition).Length > 0`.

Question now: Do we want to have this disabled for hovering helicopters (I'd say yes).
This could also be left to another PR, as it seems to be a general "bug" with aircraft.

reaperrr replied:
So, should I change the `Aircraft.IsMoving` check to only return `true` if the aircraft is moving (horizontally or vertically)?

pchote replied:
For at least the orca `WithMoveAnimation` case we'll only want to consider horizontal movements.  I believe that we've already ported all the other fiddly uses of `Aircraft.IsMoving` to upgrades so I *think* (but please check) we're otherwise safe to make that change.

abcdefg30 commented:
Looks good to me otherwise. :+1: / :white_check_mark:

reaperrr commented:
Updated and rebased.

I didn't update the testcase, instead the TD Orca now uses this instead of `WithMoveAnimation`.
Note: I have an upgrade rule ready to get rid of `WithMoveAnimation`, but we need to make `WithAttackAnimation` able to handle multiple sprite body traits before we can use this for actors that use both (like the TS Wolverine).

@abcdefg30 could you revalidate your +1?

pchote commented:
We need to complete rewrite/rethink the way `WithInfantryBody` etc work with respect to animations, the current approach just doesn't work (and changing WithAttackAnimation etc won't help).  The body trait needs to listen to these things itself so that it can handle the transitions between them (e.g. tiltrotor animations, or liedown/standup).

reaperrr commented:
Generally speaking I agree, but that sounds like another larger-scale refactor and those usually take time.

abcdefg30 commented:
Looks still good to me. 👍 / ✅

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs:
```diff
@@ -0,0 +1,75 @@
+		public readonly string[] Upgrades = { };
+
+		[Desc("Apply upgrades on straight vertical movement as well.")]
+		public readonly bool ConsiderStraightVerticalMovement = false;
```
I find this naming confusing.  Lets simplify this to `ConsiderVerticalMovement`?

pchote commented on mods/cnc/rules/aircraft.yaml:
```diff
@@ -158,8 +158,15 @@ ORCA:
-		MoveSequence: move
+	UpgradeOnMovement:
+		Upgrades: movingforward
+	WithFacingSpriteBody:
```
Given our more recent discovery / discussion about animations and sprite bodies I think it might be a better idea to keep the Orca as it is, keeping this trait for downstream modders.

pchote reviewed (changes requested):
Code changes look fine, and works as advertised so :+1: after we resolve the following.

reaperrr commented:
Updated as requested.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5e4566ddb90509af8d946ebcd1d028a8dd0a43bd
